rust/poker: Add mentoring notes (#1658)
exercism/website-copy: tracks/rust/exercises/poker/mentoring.md

### Concepts

- Iterators
- Iterator Methods
- Pattern matching

> - modelling a problem with the type system

Reviewer: ```suggestion
- Modelling a problem with the type system
```

> - Modelling a problem with the type system

Reviewer 2: ```suggestion
- Modeling a problem with the type system
```

### Reasonable Solutions

A reasonable solution should:

> - Use iterator whenever possible instead of for loops with conditionals inside

Reviewer 2: This advice always makes me skeptical. Rust's iterators are powerful and nice, but it's not justified to claim that they should always be preferred over for loops. In particular, if contained code requires a bunch of `?` operators, for loops can be much simpler to use.

Author: In general I completely agree with you. However in this case, the emphasis is on the `conditionals inside` rather than having a for loop with some ifs and complicated logic, having iterators with combinators (.filter and the like) usually makes intent much clearer (IMO)

                    (None, false) => HandRank::HighCard(*h2, *h1, *m, *l2, *l1),
                    (None, true) => HandRank::Flush(*h2, *h1, *m, *l2, *l1),
                    (Some(high), false) => HandRank::Straight(high),
                    (Some(high), true) => HandRank::StraightFlush(high),
                }
            }
            _ => unreachable!(),
        }
    }
}

/// Given a list of poker hands, return a list of those hands which win.
///
/// Note the type signature: this function should return _the same_ reference to
/// the winning hand(s) as were passed in, not reconstructed strings which happen to be equal.
pub fn winning_hands<'a>(hands: &[&'a str]) -> Option<Vec<&'a str>> {
    if hands.is_empty() {
        return None;
    }
    let best_hand = hands
        .iter()
        .cloned()
        .map(Hand::new)
        .map(HandRank::new)
        .max()
        .unwrap();
    hands
        .iter()
        .filter(|s| HandRank::new(Hand::new(*s)).cmp(&best_hand) == Ordering::Equal)
        .cloned()
        .collect::<Vec<_>>()
        .into()
}
```

### Common suggestions

#### No need to store a reference to the given string

> storing hand_str inside the Hand struct adds overhead and isn't necessary. It's better to filter on the given hands argument to return the winning hands.

Reviewer: ```suggestion
Storing `hand_str` inside the `Hand` struct adds overhead and isn't necessary. It's better to filter on the given hands argument to return the winning hands.
```

Reviewer: Perhaps you can add _why_ this is better.

#### Using complex logic to determine the hand category

> a lot of students will go into 10-20 lines logic trying to determine the number of pairs, the number of triples... That logic can usually be replaced with collecting ranks into a hashmap and pattern macthing on it.

Reviewer: ```suggestion
Many students will go into 10-20 lines logic trying to determine the number of pairs, the number of triples. That logic can usually be replaced with collecting ranks into a `HashMap` and us pattern matching on it.
```
